### PR TITLE
contrib: embed C++11 patch in install_db4.sh

### LINKED ISSUE #20722: contrib/install_db4.sh fetching a patch from gist.githubusercontent.com
Berkeley DB installer in contrib makes a call over the internet to https://gist.githubusercontent.com to fetch a 141 line patch.

https://github.com/bitcoin/bitcoin/blob/f1dbf92ff0475a01d20170ea422c1d086acbbc57/contrib/install_db4.sh#L71

The gist has [multiple revisions](
https://gist.github.com/LnL7/5153b251fd525fe15de69b67e63a6075/revisions)

**Expected behavior**

One less call over the internet during the build process. One less website to depend on.

**To reproduce**

```
 ./contrib/install_db4.sh 
```

`install_db4.sh` is recommended by the `doc/build-*.md` for example https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md#berkeley-db 

CC: @laanwj

### PR DESCRIPTION
**Pull request by gruve-p**
This is a continuation of https://github.com/bitcoin/bitcoin/pull/20665.

Closes #20722.

### CODE REVIEW

File: contrib/install_db4.sh [hunk 1]
```diff
@@ -68,10 +68,155 @@ tar -xzvf ${BDB_VERSION}.tar.gz -C "$BDB_PREFIX"
...
++++ b/mp/mp_fget.c
+@@ -617,7 +617,7 @@ alloc:		/* Allocate a new buffer header and data space. */
+
+ 		/* Initialize enough so we can call __memp_bhfree. */
+ 		alloc_bhp->flags = 0;
+-		atomic_init(&alloc_bhp->ref, 1);
++		atomic_init_db(&alloc_bhp->ref, 1);
+ #ifdef DIAGNOSTIC
+ 		if ((uintptr_t)alloc_bhp->buf & (sizeof(size_t) - 1)) {
+ 			__db_errx(env,
```
Comment (jonasschnelli) [inline contrib/install_db4.sh:130]: Where is the `DB_STR("3025",` gone?

File: contrib/install_db4.sh [hunk 2]
```diff
@@ -68,10 +68,155 @@ tar -xzvf ${BDB_VERSION}.tar.gz -C "$BDB_PREFIX"
...
+@@ -911,7 +911,7 @@ alloc:		/* Allocate a new buffer header and data space. */
+ 			MVCC_MPROTECT(bhp->buf, mfp->stat.st_pagesize,
```
Comment (jonasschnelli) [inline contrib/install_db4.sh:132]: This was `MVCC_MPROTECT(bhp->buf, mfp->pagesize,` before.

File: contrib/install_db4.sh [hunk 3]
```diff
@@ -68,10 +68,155 @@ tar -xzvf ${BDB_VERSION}.tar.gz -C "$BDB_PREFIX"
...
+ 			     MTX_MPOOL_FILE_BUCKET, 0, &htab[i].mtx_hash)) != 0)
+ 				return (ret);
+ 			SH_TAILQ_INIT(&htab[i].hash_bucket);
+-			atomic_init(&htab[i].hash_page_dirty, 0);
++			atomic_init_db(&htab[i].hash_page_dirty, 0);
+ 		}
+
+ 		/*
+@@ -269,7 +269,7 @@ __memp_init(env, dbmp, reginfo_off, htab_buckets, max_nreg)
+ 		hp->mtx_hash = (mtx_base == MUTEX_INVALID) ? MUTEX_INVALID :
```
Comment (jonasschnelli) [inline contrib/install_db4.sh:176]:
```diff
-@@ -269,7 +269,7 @@ __memp_init(env, dbmp, reginfo_off, htab_buckets, max_nreg)
- 		hp->mtx_hash = (mtx_base == MUTEX_INVALID) ? MUTEX_INVALID :
- 		    mtx_base + i;
+@@ -302,7 +302,7 @@ no_prealloc:
+ 		} else
+ 			hp->mtx_hash = mtx_base + (i % dbenv->mp_mtxcount);
```

**Review by jonasschnelli — CHANGES_REQUESTED** [review]

**Comment by gruve-p** [comment]
@jonasschnelli The patch is an exact copy of https://github.com/bitcoin/bitcoin/blob/master/depends/patches/bdb/clang_cxx_11.patch

**Comment by laanwj** [comment]
Concept ACK

**Review by fanquake — APPROVED** [review]
ACK 92370033a2606ee88f4d080193e0bf6da481dc2e. 

I checked that the patch is the same as what we carry in depends, and that the `*.a`s produced by the script are identical between master and this PR on macOS and Linux. i.e on macOS:
```bash
# master 7b975639ef93b50537a3ec6326b54d7218afc8da
CFLAGS="-Wno-error=implicit-function-declaration" ZERO_AR_DATE=1 ./contrib/install_db4.sh $(pwd)
mv db4 db4_master
# checkout #20906
CFLAGS="-Wno-error=implicit-function-declaration" ZERO_AR_DATE=1 ./contrib/install_db4.sh $(pwd)
diffoscope db4/lib db4_master/lib/ --exclude-directory-metadata=yes
```

**Comment by practicalswift** [comment]
Concept ACK: in-repo is more robust and easier to review than out-of-repo :)

**Comment by jamesob** [comment]
I am concept ACK on this, but why did you choose to change the patch format? This makes the change more difficult to review.

**Comment by gruve-p** [comment]
@jamesob The patch format is exactly the same as we use for depends: https://github.com/bitcoin/bitcoin/blob/master/depends/patches/bdb/clang_cxx_11.patch

**Comment by laanwj** [comment]
It would have been easier for reviewers if you made it clear in the PR description that this is now using the `depends/patches/bdb/clang_cxx_11.patch` and not the one from the gist.

Anyhow—I checked that it (apart from whitespace) matches.

ACK 92370033a2606ee88f4d080193e0bf6da481dc2e

**Pull request merged**